fabtests/examples/example_rdm.c

```diff
@@ -0,0 +1,507 @@
+	if (ret)
+		printf("warning: error closing CQ (%d)\n", ret);
+
+	/* fi_close(struct fid *domain); */
+	ret = fi_close(&domain->fid);
+	if (ret)
+		printf("warning: error closing domain (%d)\n", ret);
+
+	/* fi_close(struct fid *fabric); */
+	ret = fi_close(&fabric->fid);
+	if (ret)
+		printf("warning: error closing fabric (%d)\n", ret);
+
+	if (info)
+		/* fi_freeinfo(struct fi_info *info); */
+		fi_freeinfo(info);
+}
+
+/*
+ * Post a receive buffer. This call does not ensure a message has been
+ * received, just that a buffer has been passed to OFI for the next message
+ * the provider receives. Receives may be directed or undirected using the
+ * address parameter. Here, we pass in the fi_addr but note that the server
+ * has not inserted the client's address into its AV, so the address is still
+ * FI_ADDR_UNSPEC, indicating that this buffer may receive incoming data from
+ * any address. An application may set this to a real fi_addr if the buffer
+ * should only receive data from a certain peer.
+ * When posting a buffer, if the provider is not ready to process messages
+ * (because of connection initialization for example), it may return
+ * -FI_EAGAIN. This does not indicate an error, but rather that the application
+ * should try again later. This is why we almost always wrap sends and receives
+ * in a do/while. Some providers may need the application to drive progress in
+ * order to get out of the -FI_EAGAIN loop. To drive progress, the application
+ * needs to call fi_cq_read (not necessarily reading any completion entries).
+ */
+static int post_recv(void)
+{
+	int ret;
+
+	do {
+		/*
+		 * fi_recv(struct fid_ep *ep, void * buf, size_t len,
+		 *		   void *desc, fi_addr_t src_addr, void *context);
+		 */
+		ret = fi_recv(ep, buf, BUF_SIZE, NULL, fi_addr, NULL);
+		if (ret && ret != -FI_EAGAIN) {
+			printf("error posting recv buffer (%d\n", ret);
+			return ret;
+		}
+		if (ret == -FI_EAGAIN)
+			(void) fi_cq_read(cq, NULL, 0);
+	} while (ret);
+
+	return 0;
+}
+
+/*
+ * Post a send buffer. This call does not ensure a message has been sent, just
+ * that a buffer has been submitted to OFI to be sent. Unlike a receive buffer,
+ * a send needs a valid fi_addr as input to tell the provider where to send the
+ * message. Similar to the receive buffer posting porcess, when posting a send
+ * buffer, if the provider is not ready to process messages, it may return
+ * -FI_EAGAIN. This does not indicate an error, but rather that the application
+ * should try again later. Just like the receive, we drive progress with
+ * fi_cq_read if this is the case.
+ */
+static int post_send(void)
+{
+	char *msg = "Hello, server! I am the client you've been waiting for!\0";
+	int ret;
+
+	(void) snprintf(buf, BUF_SIZE, "%s", msg);
+
+	do {
+		/*
+		 * fi_send(struct fid_ep *ep, const void *buf, size_t len,
+		 *	   void *desc, fi_addr_t dest_addr,
+		 *	   void *context);
+		 */
+		ret = fi_send(ep, buf, BUF_SIZE, NULL, fi_addr, NULL);
```

Comment:
Note that this assumes an inject size >= sizeof(msg), or that memory registration isn't needed.  That should be captured in the hints (if assuming inject_size), or a comment (if we're not supporting memory registration).